packages/licensing/src/class-licensing.php

```diff
@@ -0,0 +1,219 @@
+<?php
+/**
+ * A Terms of Service class for Jetpack.
+ *
+ * @package automattic/jetpack-licensing
+ */
+
+namespace Automattic\Jetpack;
+
+use Automattic\Jetpack\Connection\Manager as Connection_Manager;
+use Jetpack;
+use Jetpack_IXR_ClientMulticall;
+use Jetpack_Options;
+use WP_Error;
+
+/**
+ * Class Licensing.
+ * Helper class that is responsible for attaching licenses to the current site.
+ *
+ * @since ??
+ */
+class Licensing {
+	/**
+	 * Name of the WordPress option that holds all known Jetpack licenses.
+	 *
+	 * @const string
+	 */
+	const LICENSES_OPTION_NAME = 'jetpack_licenses';
+
+	/**
+	 * Name of the WordPress transient that holds the last license attaching error, if any.
+	 *
+	 * @const string
+	 */
+	const ERROR_TRANSIENT_NAME = 'jetpack_licenses_error';
+
+	/**
+	 * Holds the singleton instance of this class.
+	 *
+	 * @var self
+	 */
+	protected static $instance = false;
+
+	/**
+	 * Singleton.
+	 *
+	 * @static
+	 */
+	public static function instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Initialize.
+	 *
+	 * @return void
+	 */
+	public function initialize() {
+		add_action( 'update_option_' . self::LICENSES_OPTION_NAME, array( $this, 'attach_stored_licenses' ) );
+		add_action( 'jetpack_authorize_ending_authorized', array( $this, 'attach_stored_licenses_on_connection' ) );
+		add_action( 'load-toplevel_page_jetpack', array( $this, 'surface_error' ) );
+	}
+
+	/**
+	 * Get Jetpack connection manager instance.
+	 *
+	 * @return Connection_Manager
+	 */
+	protected function connection() {
+		return new Connection_Manager();
+	}
+
+	/**
+	 * Make an authenticated WP.com API multicall request instance.
+	 *
+	 * @param array $client_args IXR client arguments.
+	 * @return Jetpack_IXR_ClientMulticall
+	 */
+	protected function request( $client_args ) {
+		return new Jetpack_IXR_ClientMulticall( $client_args );
+	}
+
+	/**
+	 * Log an error to be surfaced to the user at a later time.
+	 *
+	 * @param string $error Human-readable error message.
+	 * @return void
+	 */
+	protected function log_error( $error ) {
+		set_transient( self::ERROR_TRANSIENT_NAME, $error );
+	}
+
+	/**
+	 * Surface a previously logged error (if any) to the user in the form of an admin notice.
+	 *
+	 * @return void
+	 */
+	public function surface_error() {
+		$error = get_transient( self::ERROR_TRANSIENT_NAME );
+		delete_transient( self::ERROR_TRANSIENT_NAME );
+
+		if ( $error ) {
+			Jetpack::state( 'error', $error );
```

Comment:
Using `Jetpack` class in the package makes it impossible to use it without Jetpack.
For example, `WCPay` relies on `jetpack-connection`, but can perfectly function without Jetpack installed. Would the `jetpack-licensing` package have potential value when it's used without Jetpack?

Reply:
The first few iterations of the Licensing project will be exclusively aimed at Jetpack plans so it being coupled to Jetpack is not necessarily a concern right now.
Regardless, would substituting it with an action which is then hooked into from Jetpack in order to run `self::state()` suffice as a solution?

Reply:
In that case we can leave it as is.
No need to jump the gun, we can address it later should the need arise.